### PR TITLE
[SPARK-10186][SQL] support postgre array type in JDBCRDD

### PR DESCRIPTION
**Pull request by cloud-fan**
Add ARRAY support to `PostgresDialect`.

Nested ARRAY is not allowed for now because it's hard to get the array dimension info. See http://stackoverflow.com/questions/16619113/how-to-get-array-base-type-in-postgres-via-jdbc

Thanks for the initial work from @mariusvniekerk !

Close https://github.com/apache/spark/pull/9137

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @marmbrus

**Comment by mariusvniekerk** [comment]
I've added write support in #9137 as well if you want to just use it from there.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 1]
```diff
@@ -121,6 +145,12 @@ object JdbcUtils extends Logging {
                 case TimestampType => stmt.setTimestamp(i + 1, row.getAs[java.sql.Timestamp](i))
                 case DateType => stmt.setDate(i + 1, row.getAs[java.sql.Date](i))
                 case t: DecimalType => stmt.setBigDecimal(i + 1, row.getDecimal(i))
+                case ArrayType(et, _) =>
+                  assert(jdbcTypes(i).databaseTypeDefinition.endsWith("[]"))
```
Comment (mariusvniekerk) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Is that the same in all backends that support arrays (Oracle etc)?

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: you are right, it's not always working

**Comment by mariusvniekerk** [comment]
These test failures don't seem to be related?

**Comment by cloud-fan** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala [hunk 2]
```diff
@@ -64,13 +64,18 @@ abstract class JdbcDialect {
...
    * @return The actual DataType (subclasses of [[org.apache.spark.sql.types.DataType]])
    *         or null if the default type mapping should be used.
    */
   def getCatalystType(
-    sqlType: Int, typeName: String, size: Int, md: MetadataBuilder): Option[DataType] = None
+    sqlType: Int,
+    typeName: String,
+    precision: Int,
+    scale: Int,
+    md: MetadataBuilder): Option[DataType] = None
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala]: This is a public API, if we add fields here we are going to break any existing implementations.  Do we have to do this?

File: docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/PostgresIntegrationSuite.scala [hunk 3]
```diff
@@ -72,11 +78,17 @@ class PostgresIntegrationSuite extends DockerJDBCIntegrationSuite {
...
   }
 
   test("Basic write test") {
     val df = sqlContext.read.jdbc(jdbcUrl, "bar", new Properties)
-    df.write.jdbc(jdbcUrl, "public.barcopy", new Properties)
     // Test only that it doesn't crash.
+    df.write.jdbc(jdbcUrl, "public.barcopy", new Properties)
+    // Test write null values.
+    df.select(df.queryExecution.analyzed.output.map { a =>
+      Column(If(Literal(true), Literal(null), a)).as(a.name)
```
Comment (marmbrus) [inline docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/PostgresIntegrationSuite.scala]: Is this just `Literal.create(null, a.dataType)`?

**Comment by marmbrus** [comment]
Thanks, I'm going to merge this into master and 1.6.

File: docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/PostgresIntegrationSuite.scala (hunk 3 above)
Comment (cloud-fan) [inline docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/PostgresIntegrationSuite.scala]: ah yea, we can simply this.

**Pull request closed without merging**